phoenix-core/src/main/java/org/apache/phoenix/coprocessor/GroupedAggregateRegionObserver.java

```diff
@@ -398,66 +414,84 @@ private RegionScanner scanUnordered(ObserverContext<RegionCoprocessorEnvironment
-                        "Spillable groupby enabled: " + spillableEnabled,
-                        ScanUtil.getCustomAnnotations(scan)));
-            }
-            Region region = c.getEnvironment().getRegion();
-            boolean acquiredLock = false;
-            try {
-                region.startRegionOperation();
-                acquiredLock = true;
-                synchronized (scanner) {
-                    do {
-                        List<Cell> results = useQualifierAsIndex ? new EncodedColumnQualiferCellsList(minMaxQualifiers.getFirst(), minMaxQualifiers.getSecond(), encodingScheme) : new ArrayList<Cell>();
-                        // Results are potentially returned even when the return
-                        // value of s.next is false
-                        // since this is an indication of whether or not there are
-                        // more values after the
-                        // ones returned
-                        hasMore = scanner.nextRaw(results);
-                        if (!results.isEmpty()) {
-                            result.setKeyValues(results);
-                            ImmutableBytesPtr key =
-                                TupleUtil.getConcatenatedValue(result, expressions);
-                            Aggregator[] rowAggregators = groupByCache.cache(key);
-                            // Aggregate values here
-                            aggregators.aggregate(rowAggregators, result);
+        return new BaseRegionScanner(scanner) {
+            RegionScanner regionScanner = null;
+            @Override
+            public boolean next(List<Cell> resultsToReturn) throws IOException {
+                boolean hasMore;
+                long startTime = EnvironmentEdgeManager.currentTimeMillis();
+                long now;
+                Tuple result = useQualifierAsIndex ? new PositionBasedMultiKeyValueTuple() : new MultiKeyValueTuple();
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(LogUtil.addCustomAnnotations(
+                            "Spillable groupby enabled: " + spillableEnabled,
+                            ScanUtil.getCustomAnnotations(scan)));
+                }
+                Region region = c.getEnvironment().getRegion();
+                boolean acquiredLock = false;
+                try {
+                    region.startRegionOperation();
+                    acquiredLock = true;
+                    synchronized (scanner) {
+                        if (regionScanner != null) {
+                            return regionScanner.next(resultsToReturn);
                         }
-                    } while (hasMore && groupByCache.size() < limit);
+                        do {
+                            List<Cell> results = useQualifierAsIndex ? new EncodedColumnQualiferCellsList(minMaxQualifiers.getFirst(), minMaxQualifiers.getSecond(), encodingScheme) : new ArrayList<Cell>();
```

Comment:
tiny nit: good to stick in a newline in there, looks like more than 100 chars wide. (If it's not, feel free to disregard)